File: LIM PO KEAT FINAL YEAR PROJECT/qr_virtual_identity/inspect_firestore.js
// inspect_firestore.js - READ-ONLY Firestore structure inspector
// Run with: node inspect_firestore.js

const http = require("http");

const FIRESTORE_EMULATOR_HOST = "127.0.0.1";
const FIRESTORE_EMULATOR_PORT = 8080;
const PROJECT_ID = "po-keat-fyp";

// Expected collections based on codebase analysis
const expectedCollections = [
  "admins",
  "users",
  "scan_points",
  "interactions",
  "logs",
  "events",
  "user_tickets",
  "guest_users",
  "guest_tickets",
  "books",
  "book_loans",
  "library_sessions",
  "ticket_scans",
  "scanner_triggers",
  "scanner_status",
];

console.log("🔍 FIRESTORE STRUCTURE INSPECTOR (READ-ONLY MODE)");
console.log("═══════════════════════════════════════════════════════════════");
console.log(
  `📡 Connecting to Firestore Emulator at ${FIRESTORE_EMULATOR_HOST}:${FIRESTORE_EMULATOR_PORT}`
);
console.log(`📋 Project ID: ${PROJECT_ID}\n`);

async function httpGet(path) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: FIRESTORE_EMULATOR_HOST,
      port: FIRESTORE_EMULATOR_PORT,
      path: path,
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    };

    const req = http.request(options, (res) => {
      let data = "";

      res.on("data", (chunk) => {
        data += chunk;
      });

      res.on("end", () => {
        try {
          const parsed = JSON.parse(data);
          resolve(parsed);
        } catch (e) {
          resolve({ error: "Failed to parse response", data });
        }
      });
    });

    req.on("error", (error) => {
      reject(error);
    });

    req.end();
  });
}

function inferSchema(data) {
  const schema = {};

  for (const [key, value] of Object.entries(data)) {
    if (value === null) {
      schema[key] = "null";
    } else if (typeof value === "string") {
      schema[key] = "string";
    } else if (typeof value === "number") {
      schema[key] = Number.isInteger(value) ? "integer" : "double";
    } else if (typeof value === "boolean") {
      schema[key] = "boolean";
    } else if (Array.isArray(value)) {
      if (value.length === 0) {
        schema[key] = "array (empty)";
      } else {
        const firstItemType = typeof value[0];
        schema[key] = `array<${firstItemType}>`;
      }
    } else if (typeof value === "object") {
      if (value._seconds !== undefined) {
        schema[key] = "Timestamp";
      } else {
        schema[key] = "map";
      }
    } else {
      schema[key] = typeof value;
    }
  }

  return schema;
}

function checkFieldConsistency(documents) {
  const allFields = new Set();
  const fieldCounts = {};

  for (const doc of documents) {
    for (const key of Object.keys(doc)) {
      allFields.add(key);
      fieldCounts[key] = (fieldCounts[key] || 0) + 1;
    }
  }

  const inconsistentFields = [];
  for (const field of allFields) {
    if (fieldCounts[field] < documents.length) {
      inconsistentFields.push(
        `${field} (${fieldCounts[field]}/${documents.length} docs)`
      );
    }
  }

  return inconsistentFields;
}

async function analyzeCollection(collectionName) {
  try {
    const path = `/v1/projects/${PROJECT_ID}/databases/(default)/documents/${collectionName}`;
    const response = await httpGet(path);

    if (response.error || !response.documents) {
      return {
        name: collectionName,
        empty: true,
        error: response.error,
      };
    }

    const documents = response.documents.map((doc) => {
      const data = {};
      if (doc.fields) {
        for (const [key, value] of Object.entries(doc.fields)) {
          // Extract the actual value based on Firestore type
          if (value.stringValue !== undefined) data[key] = value.stringValue;
          else if (value.integerValue !== undefined)
            data[key] = parseInt(value.integerValue);
          else if (value.doubleValue !== undefined)
            data[key] = value.doubleValue;
          else if (value.booleanValue !== undefined)
            data[key] = value.booleanValue;
          else if (value.timestampValue !== undefined)
            data[key] = { _seconds: value.timestampValue };
          else if (value.arrayValue !== undefined)
            data[key] = value.arrayValue.values || [];
          else if (value.mapValue !== undefined)
            data[key] = value.mapValue.fields || {};
          else if (value.nullValue !== undefined) data[key] = null;
          else data[key] = value;
        }
      }
      return {
        id: doc.name.split("/").pop(),
        data,
      };
    });

    return {
      name: collectionName,
      empty: false,
      count: documents.length,
      sampleIds: documents.slice(0, 5).map((d) => d.id),
      schema: documents.length > 0 ? inferSchema(documents[0].data) : {},
      inconsistentFields:
        documents.length > 1
          ? checkFieldConsistency(documents.map((d) => d.data))
          : [],
    };
  } catch (error) {
    return {
      name: collectionName,
      error: error.message,
    };
  }
}

async function main() {
  console.log("📋 ROOT COLLECTION LIST");
  console.log(
    "═══════════════════════════════════════════════════════════════\n"
  );

  const results = [];

  for (const collectionName of expectedCollections) {
    const result = await analyzeCollection(collectionName);
    results.push(result);

    if (result.error) {
      console.log(`❌ ${collectionName} (error: ${result.error})`);
    } else if (result.empty) {
      console.log(`⚪ ${collectionName} (empty)`);
    } else {
      console.log(`✅ ${collectionName} (${result.count} documents)`);
    }
  }

  console.log("\n");
  console.log(
    "═══════════════════════════════════════════════════════════════"
  );
  console.log("📊 DETAILED COLLECTION ANALYSIS");
  console.log(
    "═══════════════════════════════════════════════════════════════\n"
  );

  for (const result of results) {
    if (result.empty || result.error) continue;

    console.log(
      "─────────────────────────────────────────────────────────────────"
    );
    console.log(`Collection: ${result.name}`);
    console.log(
      "─────────────────────────────────────────────────────────────────"
    );
    console.log(`Sample Document IDs: ${result.sampleIds.join(", ")}`);
    console.log(`Documents Sampled: ${result.count}`);
    console.log("\nInferred Schema (from first document):");
    console.log(JSON.stringify(result.schema, null, 2));

    console.log("\nField Consistency Check:");
    if (result.inconsistentFields.length === 0) {
      console.log("✅ All fields consistent across sampled documents");
    } else {
      console.log("⚠️  Inconsistent fields detected:");
      for (const field of result.inconsistentFields) {
        console.log(`   - ${field}`);
      }
    }
    console.log("");
  }

  console.log(
    "═══════════════════════════════════════════════════════════════"
  );
  console.log("📈 FIRESTORE DATABASE SUMMARY");
  console.log(
    "═══════════════════════════════════════════════════════════════\n"
  );

  const populated = results.filter((r) => !r.empty && !r.error);
  const empty = results.filter((r) => r.empty);
  const errors = results.filter((r) => r.error);

  console.log(`Total Collections Checked: ${expectedCollections.length}`);
  console.log(`Collections with Data: ${populated.length}`);
  console.log(`Empty Collections: ${empty.length}`);
  console.log(`Errors: ${errors.length}`);

  if (empty.length > 0) {
    console.log("\n⚠️  Empty Collections:");
    for (const result of empty) {
      console.log(`   - ${result.name}`);
    }
  }

  if (errors.length > 0) {
    console.log("\n❌ Collections with Errors:");
    for (const result of errors) {
      console.log(`   - ${result.name}: ${result.error}`);
    }
  }

  // Summary analysis
  console.log("\n🔍 DESIGN ANALYSIS:");
  console.log(
    "─────────────────────────────────────────────────────────────────"
  );

  if (populated.some((r) => r.name === "scan_points")) {
    console.log(
      '✅ Using modern "scan_points" collection (replaces "merchants")'
    );
  }

  if (populated.some((r) => r.name === "interactions")) {
    console.log(
      '✅ Using modern "interactions" collection (replaces "transactions")'
    );
  }

  if (
    populated.some((r) => r.name === "guest_users") &&
    populated.some((r) => r.name === "guest_tickets")
  ) {
    console.log("✅ Guest event ticket system implemented");
  }

  if (
    populated.some((r) => r.name === "books") &&
    populated.some((r) => r.name === "book_loans")
  ) {
    console.log("✅ Library management system with books and loans");
  }

  if (
    populated.some((r) => r.name === "scanner_triggers") &&
    populated.some((r) => r.name === "scanner_status")
  ) {
    console.log("✅ Cross-device scanner synchronization system present");
  }

  console.log("\n✅ Inspection complete!\n");
}

main().catch(console.error);

File: LIM PO KEAT FINAL YEAR PROJECT/qr_virtual_identity/inspect_firestore.dart
// inspect_firestore.dart - READ-ONLY Firestore structure inspector
import 'dart:convert';
import 'package:firebase_core/firebase_core.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'lib/firebase_options.dart';

Future<void> main() async {
  print('🔍 FIRESTORE STRUCTURE INSPECTOR (READ-ONLY MODE)');
  print('═══════════════════════════════════════════════════════════════');

  // Initialize Firebase
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);

  // Connect to Firestore emulator
  FirebaseFirestore.instance.useFirestoreEmulator('127.0.0.1', 8080);
  print('✅ Connected to Firestore Emulator at 127.0.0.1:8080\n');

  final db = FirebaseFirestore.instance;

  // List of expected collections based on codebase analysis
  final expectedCollections = [
    'admins',
    'users',
    'scan_points',
    'interactions',
    'logs',
    'events',
    'user_tickets',
    'guest_users',
    'guest_tickets',
    'books',
    'book_loans',
    'library_sessions',
    'ticket_scans',
    'scanner_triggers',
    'scanner_status',
  ];

  print('📋 ROOT COLLECTION LIST');
  print('═══════════════════════════════════════════════════════════════\n');

  final foundCollections = <String>[];
  final emptyCollections = <String>[];

  for (final collectionName in expectedCollections) {
    try {
      final snapshot = await db.collection(collectionName).limit(5).get();

      if (snapshot.docs.isEmpty) {
        emptyCollections.add(collectionName);
        print('⚪ $collectionName (empty)');
      } else {
        foundCollections.add(collectionName);
        print('✅ $collectionName (${snapshot.docs.length}+ documents)');
      }
    } catch (e) {
      print('❌ $collectionName (error: $e)');
    }
  }

  print('\n');
  print('═══════════════════════════════════════════════════════════════');
  print('📊 DETAILED COLLECTION ANALYSIS');
  print('═══════════════════════════════════════════════════════════════\n');

  // Analyze each found collection
  for (final collectionName in foundCollections) {
    await analyzeCollection(db, collectionName);
    print('');
  }

  // Summary
  print('═══════════════════════════════════════════════════════════════');
  print('📈 FIRESTORE DATABASE SUMMARY');
  print('═══════════════════════════════════════════════════════════════\n');

  print('Total Collections Checked: ${expectedCollections.length}');
  print('Collections with Data: ${foundCollections.length}');
  print('Empty Collections: ${emptyCollections.length}');

  if (emptyCollections.isNotEmpty) {
    print('\n⚠️  Empty Collections:');
    for (final name in emptyCollections) {
      print('   - $name');
    }
  }

  print('\n✅ Inspection complete!\n');
}

Future<void> analyzeCollection(
  FirebaseFirestore db,
  String collectionName,
) async {
  print('─────────────────────────────────────────────────────────────────');
  print('Collection: $collectionName');
  print('─────────────────────────────────────────────────────────────────');

  try {
    // Get first 10 documents for analysis
    final snapshot = await db.collection(collectionName).limit(10).get();
    final docCount = snapshot.docs.length;

    // Sample document IDs
    final sampleIds = snapshot.docs.take(5).map((doc) => doc.id).toList();
    print('Sample Document IDs: ${sampleIds.join(", ")}');
    print('Documents Sampled: $docCount');

    if (snapshot.docs.isEmpty) {
      print('Status: Empty collection\n');
      return;
    }

    // Infer schema from first document
    final firstDoc = snapshot.docs.first;
    final data = firstDoc.data();

    print('\nInferred Schema (from first document):');
    final schema = inferSchema(data);
    print(const JsonEncoder.withIndent('  ').convert(schema));

    // Check for subcollections (only check first document)
    print('\nChecking for subcollections...');
    final subcollections = await checkSubcollections(firstDoc.reference);

    if (subcollections.isNotEmpty) {
      print('Subcollections Found:');
      for (final subName in subcollections) {
        print('  - $subName');
      }
    } else {
      print('No subcollections found');
    }

    // Field consistency check across multiple documents
    if (snapshot.docs.length > 1) {
      print('\nField Consistency Check:');
      final allFields = <String>{};
      final fieldCounts = <String, int>{};

      for (final doc in snapshot.docs) {
        final docData = doc.data();
        for (final key in docData.keys) {
          allFields.add(key);
          fieldCounts[key] = (fieldCounts[key] ?? 0) + 1;
        }
      }

      final inconsistentFields = <String>[];
      for (final field in allFields) {
        if (fieldCounts[field]! < snapshot.docs.length) {
          inconsistentFields.add(
            '$field (${fieldCounts[field]}/${snapshot.docs.length} docs)',
          );
        }
      }

      if (inconsistentFields.isEmpty) {
        print('✅ All fields consistent across sampled documents');
      } else {
        print('⚠️  Inconsistent fields detected:');
        for (final field in inconsistentFields) {
          print('   - $field');
        }
      }
    }
  } catch (e) {
    print('❌ Error analyzing collection: $e');
  }
}

Map<String, dynamic> inferSchema(Map<String, dynamic> data) {
  final schema = <String, dynamic>{};

  for (final entry in data.entries) {
    final value = entry.value;

    if (value == null) {
      schema[entry.key] = 'null';
    } else if (value is String) {
      schema[entry.key] = 'string';
    } else if (value is int) {
      schema[entry.key] = 'integer';
    } else if (value is double) {
      schema[entry.key] = 'double';
    } else if (value is bool) {
      schema[entry.key] = 'boolean';
    } else if (value is Timestamp) {
      schema[entry.key] = 'Timestamp';
    } else if (value is List) {
      if (value.isEmpty) {
        schema[entry.key] = 'array (empty)';
      } else {
        final firstItemType = value.first.runtimeType.toString();
        schema[entry.key] = 'array<$firstItemType>';
      }
    } else if (value is Map) {
      schema[entry.key] = 'map';
    } else {
      schema[entry.key] = value.runtimeType.toString();
    }
  }

  return schema;
}

Future<List<String>> checkSubcollections(DocumentReference docRef) async {
  // Note: Firestore doesn't have a direct API to list subcollections in client SDK
  // We can only check known subcollection names
  final knownSubcollections = ['joined_events', 'transactions', 'activities'];
  final found = <String>[];

  for (final subName in knownSubcollections) {
    try {
      final subSnap = await docRef.collection(subName).limit(1).get();
      if (subSnap.docs.isNotEmpty) {
        found.add(subName);
      }
    } catch (e) {
      // Subcollection doesn't exist or error
    }
  }

  return found;
}
